dbms/src/Processors/Formats/Impl/RegexpRowInputFormat.cpp

```diff
@@ -0,0 +1,190 @@
+#include <stdlib.h>
+#include <common/find_symbols.h>
+#include <Processors/Formats/Impl/RegexpRowInputFormat.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <IO/ReadHelpers.h>
+#include <re2/stringpiece.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_DATA;
+    extern const int BAD_ARGUMENTS;
+}
+
+RegexpRowInputFormat::RegexpRowInputFormat(
+        ReadBuffer & in_, const Block & header_, Params params_, const FormatSettings & format_settings_)
+        : IRowInputFormat(header_, in_, std::move(params_)), buf(in_), format_settings(format_settings_), regexp(format_settings_.regexp.regexp)
+{
+    size_t fields_count = regexp.NumberOfCapturingGroups();
+    matched_fields.resize(fields_count);
+    re2_arguments.resize(fields_count);
+    re2_arguments_ptrs.resize(fields_count);
+    for (size_t i = 0; i != fields_count; ++i)
+    {
+        // Bind an argument to a matched field.
+        re2_arguments[i] = &matched_fields[i];
+        // Save pointer to argument.
+        re2_arguments_ptrs[i] = &re2_arguments[i];
+    }
+
+    field_format = stringToFormat(format_settings_.regexp.escaping_rule);
+}
+
+RegexpRowInputFormat::FieldFormat RegexpRowInputFormat::stringToFormat(const String & format)
+{
+    if (format == "Escaped")
+        return FieldFormat::Escaped;
+    if (format == "Quoted")
+        return FieldFormat::Quoted;
+    if (format == "CSV")
+        return FieldFormat::Csv;
+    if (format == "JSON")
+        return FieldFormat::Json;
+    throw Exception("Unknown field format \"" + format + "\".", ErrorCodes::BAD_ARGUMENTS);
+}
+
+bool RegexpRowInputFormat::readField(size_t index, MutableColumns & columns)
+{
+    const auto & type = getPort().getHeader().getByPosition(index).type;
+    bool parse_as_nullable = format_settings.null_as_default && !type->isNullable();
+    bool read = true;
+    ReadBuffer field_buf(matched_fields[index].data(), matched_fields[index].size(), 0);
+    try
+    {
+        switch (field_format)
+        {
+            case FieldFormat::Escaped:
+                if (parse_as_nullable)
+                    read = DataTypeNullable::deserializeTextEscaped(*columns[index], field_buf, format_settings, type);
```

Comment:
We can do it a more simple way...
If `parse_as_nullable`, wrap `type` with `makeNullable`.
Then parse. And if the original type was not Nullable, check parsed value for NULL.

Reply:
@tavplubix I see similar code in Template format. Why we didn't implement it in that way?

Reply:
As I remember, it was previously implemented that way in `CSVRowInputFormat`, but it was slower because of additional check for NULL and copying from `ColumnNullable` to original column.

Reply:
Ok.